File: FBDeviceControlTests/Tests/Unit/FBAMRestorableDeviceTests.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import <XCTest/XCTest.h>

#import <FBControlCore/FBControlCore.h>
#import <FBDeviceControl/FBAMRestorableDevice.h>
#import <FBDeviceControl/FBDeviceCommands.h>

@interface FBAMRestorableDeviceTests : XCTestCase

@end

@implementation FBAMRestorableDeviceTests

- (FBAMRestorableDevice *)deviceWithAllValues:(NSDictionary<NSString *, id> *)allValues
{
  AMDCalls calls = {};
  return [[FBAMRestorableDevice alloc]
          initWithCalls:calls
          restorableDevice:(__bridge AMRestorableDeviceRef) @"fake-restorable-ref"
          allValues:allValues
          workQueue:dispatch_queue_create("com.facebook.fbdevicecontrol.tests.work", DISPATCH_QUEUE_SERIAL)
          asyncQueue:dispatch_queue_create("com.facebook.fbdevicecontrol.tests.async", DISPATCH_QUEUE_SERIAL)
          logger:FBControlCoreGlobalConfiguration.defaultLogger];
}

- (void)testStringProductTypeFlowsIntoDeviceType
{
  FBAMRestorableDevice *device = [self deviceWithAllValues:@{FBDeviceKeyProductType : @"iPhone14,2"}];
  XCTAssertEqualObjects(device.deviceType.model, @"iPhone14,2");
}

- (void)testStringDeviceNameFlowsIntoName
{
  FBAMRestorableDevice *device = [self deviceWithAllValues:@{FBDeviceKeyDeviceName : @"lab-device"}];
  XCTAssertEqualObjects(device.name, @"lab-device");
}

- (void)testNumericUniqueChipIDFlowsIntoUniqueIdentifier
{
  FBAMRestorableDevice *device = [self deviceWithAllValues:@{FBDeviceKeyUniqueChipID : @12345}];
  XCTAssertEqualObjects(device.uniqueIdentifier, @"12345");
}

- (void)testMissingDeviceNameFallsBackToUnknown
{
  FBAMRestorableDevice *device = [self deviceWithAllValues:@{}];
  XCTAssertEqualObjects(device.name, @"unknown");
}

- (void)testMissingUniqueChipIDFallsBackToUnknown
{
  FBAMRestorableDevice *device = [self deviceWithAllValues:@{}];
  XCTAssertEqualObjects(device.uniqueIdentifier, @"unknown");
}

@end

File: FBDeviceControlTests/Tests/Unit/FBAMRestorableDeviceTests.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import FBControlCore
import FBDeviceControl
import Foundation
import Testing

/// The restorable device's tolerant reads: restore info dictionaries are populated by MobileDevice
/// and may omit keys or carry unexpectedly-typed values, so every accessor must degrade to the
/// unknown value rather than trap.
@Suite
struct FBAMRestorableDeviceTests {

  private func device(allValues: [String: Any]) -> FBAMRestorableDevice {
    FBAMRestorableDevice(
      calls: FBCreateZeroedAMDCalls(),
      restorableDevice: "fake-restorable-ref" as AnyObject,
      allValues: allValues,
      work: DispatchQueue(label: "com.facebook.fbdevicecontrol.tests.work"),
      asyncQueue: DispatchQueue(label: "com.facebook.fbdevicecontrol.tests.async"),
      logger: FBControlCoreGlobalConfiguration.defaultLogger)
  }

  @Test
  func stringProductTypeFlowsIntoDeviceType() {
    #expect(device(allValues: [FBDeviceKey.productType.rawValue: "iPhone14,2"]).deviceType.model.rawValue == "iPhone14,2")
  }

  @Test
  func stringDeviceNameFlowsIntoName() {
    #expect(device(allValues: [FBDeviceKey.deviceName.rawValue: "lab-device"]).name == "lab-device")
  }

  @Test
  func numericUniqueChipIDFlowsIntoUniqueIdentifier() {
    #expect(device(allValues: [FBDeviceKey.uniqueChipID.rawValue: 12345]).uniqueIdentifier == "12345")
  }

  @Test
  func missingDeviceNameFallsBackToUnknown() {
    #expect(device(allValues: [:]).name == "unknown")
  }

  @Test
  func missingUniqueChipIDFallsBackToUnknown() {
    #expect(device(allValues: [:]).uniqueIdentifier == "unknown")
  }

  @Test
  func missingProductTypeFallsBackToUnknownDeviceType() {
    #expect(device(allValues: [:]).deviceType.model.rawValue == "unknown")
  }

  @Test
  func nonStringProductTypeFallsBackToUnknownDeviceType() {
    #expect(device(allValues: [FBDeviceKey.productType.rawValue: 42]).deviceType.model.rawValue == "unknown")
  }

  @Test
  func stringUniqueChipIDFlowsIntoUniqueIdentifier() {
    #expect(device(allValues: [FBDeviceKey.uniqueChipID.rawValue: "chip-as-string"]).uniqueIdentifier == "chip-as-string")
  }

  @Test
  func nonStringDeviceNameFallsBackToUnknown() {
    #expect(device(allValues: [FBDeviceKey.deviceName.rawValue: 99]).name == "unknown")
  }
}
